File: src/main/java/com/github/cybellereaper/wizpets/api/talent/PetTalent.java
package com.github.cybellereaper.wizpets.api.talent;

import com.github.cybellereaper.wizpets.api.ActivePet;
import com.github.cybellereaper.wizpets.api.StatType;
import org.bukkit.entity.LivingEntity;

/**
 * Represents a behaviour module that can be attached to an active pet.
 */
public interface PetTalent {
    String getId();

    String getDisplayName();

    String getDescription();

    default void onSummon(ActivePet pet) {
    }

    default void tick(ActivePet pet) {
    }

    default void onAttack(ActivePet pet, LivingEntity target, double baseDamage) {
    }

    default void onDismiss(ActivePet pet) {
    }

    default double modifyStat(ActivePet pet, StatType stat, double current) {
        return current;
    }
}

File: src/main/kotlin/com/github/cybellereaper/wizpets/api/talent/PetTalent.kt
package com.github.cybellereaper.wizpets.api.talent

import com.github.cybellereaper.wizpets.api.ActivePet
import com.github.cybellereaper.wizpets.api.StatType
import org.bukkit.entity.LivingEntity

/**
 * Represents a behaviour module that can be attached to an active pet.
 */
interface PetTalent {
    val id: String
    val displayName: String
    val description: String

    fun onSummon(pet: ActivePet) {}
    fun tick(pet: ActivePet) {}
    fun onAttack(pet: ActivePet, target: LivingEntity, baseDamage: Double) {}
    fun onDismiss(pet: ActivePet) {}

    /**
     * Allows talents to influence stats before calculations are finalised.
     */
    fun modifyStat(pet: ActivePet, stat: StatType, current: Double): Double = current
}

fun interface TalentFactory {
    fun create(): PetTalent
}
